File: moqt/session_stream.go
package moqt

import (
	"context"
	"fmt"
	"log/slog"
	"slices"
	"sync"

	"github.com/OkutaniDaichi0106/gomoqt/moqt/bitrate"
	"github.com/OkutaniDaichi0106/gomoqt/moqt/internal/message"
	"github.com/OkutaniDaichi0106/gomoqt/quic"
)

func newSessionStream(stream quic.Stream, req *SetupRequest, detector bitrate.ShiftDetector) *sessionStream {
	ss := &sessionStream{
		ctx:          context.WithValue(stream.Context(), &biStreamTypeCtxKey, message.StreamTypeSession),
		stream:       stream,
		Version:      DefaultServerVersion, // Default version before setup
		SetupRequest: req,
		updatedCh:    make(chan struct{}, 1),
		detector:     detector,
	}
	return ss
}

type sessionStream struct {
	ctx       context.Context
	updatedCh chan struct{}

	localBitrate  uint64 // The bitrate set by the local
	remoteBitrate uint64 // The bitrate set by the remote

	stream quic.Stream

	mu sync.Mutex

	// Version of the protocol used in this session
	Version Version

	// Setup request from the client
	*SetupRequest

	// Parameters specified by the server
	ServerExtensions *Extension

	// Detector of significant BPS changes
	detector bitrate.ShiftDetector

	listenOnce sync.Once
}

func newResponse(sessStr *sessionStream) *response {
	return &response{
		sessionStream: sessStr,
	}
}

type response struct {
	*sessionStream
	onceSetup sync.Once
}

func (r *response) AwaitAccepted() error {
	var err error
	r.onceSetup.Do(func() {
		var sum message.SessionServerMessage
		err = sum.Decode(r.stream)
		if err != nil {
			return
		}
		r.Version = Version(sum.SelectedVersion)
		r.ServerExtensions = &Extension{sum.Parameters}

		r.handleUpdates()
	})

	return err
}

var _ SetupResponseWriter = (*responseWriter)(nil)

func newResponseWriter(conn quic.Connection, sessStr *sessionStream, connLogger *slog.Logger, server *Server) *responseWriter {
	return &responseWriter{
		sessionStream: sessStr,
		conn:          conn,
		connLogger:    connLogger,
		server:        server,
	}
}

type responseWriter struct {
	*sessionStream
	conn       quic.Connection
	connLogger *slog.Logger
	server     *Server
	onceSetup  sync.Once
}

func (w *responseWriter) SelectVersion(v Version) error {
	if !slices.Contains(w.Versions, v) {
		return fmt.Errorf("version %d not supported by client", v)
	}
	w.Version = v
	return nil
}

func (w *responseWriter) SetExtensions(extensions *Extension) {
	w.ServerExtensions = extensions
}

func (w *responseWriter) accept(mux *TrackMux) (*Session, error) {
	var err error
	w.onceSetup.Do(func() {
		// TODO: Implement setup logic if needed
		var params parameters
		if w.ServerExtensions != nil {
			params = w.ServerExtensions.parameters
		}
		err = message.SessionServerMessage{
			SelectedVersion: uint64(w.Version),
			Parameters:      params,
		}.Encode(w.stream)
		if err != nil {
			return
		}

		// Start listening for updates
		w.handleUpdates()
	})

	if err != nil {
		return nil, err
	}

	var sess *Session
	sess = newSession(w.conn, w.sessionStream, mux, w.connLogger, func() { w.server.removeSession(sess) })
	w.server.addSession(sess)

	return sess, nil
}

func (w *responseWriter) Reject(code SessionErrorCode) error {
	w.stream.CancelWrite(quic.StreamErrorCode(code))
	w.stream.CancelRead(quic.StreamErrorCode(code))
	return nil
}

func (ss *sessionStream) updateSession(bitrate uint64) error {
	ss.mu.Lock()
	defer ss.mu.Unlock()

	err := message.SessionUpdateMessage{
		Bitrate: bitrate,
	}.Encode(ss.stream)
	if err != nil {
		return Cause(ss.ctx)
	}

	ss.localBitrate = bitrate

	return nil
}

// handleUpdates triggers the goroutine to start listening for session updates
func (ss *sessionStream) handleUpdates() {
	// Safe to call multiple times
	ss.listenOnce.Do(func() {
		go func() {
			var sum message.SessionUpdateMessage
			var err error

			for {
				err = sum.Decode(ss.stream)
				if err != nil {
					break
				}

				ss.mu.Lock()
				ss.remoteBitrate = sum.Bitrate
				select {
				case ss.updatedCh <- struct{}{}:
				default:
				}
				ss.mu.Unlock()
			}

			ss.mu.Lock()
			if ss.updatedCh != nil {
				close(ss.updatedCh)
			}
			ss.mu.Unlock()
		}()
	})
}

func (ss *sessionStream) SessionUpdated() <-chan struct{} {
	ss.mu.Lock()
	defer ss.mu.Unlock()
	return ss.updatedCh
}

func (ss *sessionStream) Context() context.Context {
	return ss.ctx
}

// Accept accepts a setup request and converts it to an active Session.
// The function expects a SetupResponseWriter as provided by the server when
// responding to a client SETUP request. It uses the provided TrackMux to
// route tracks for the accepted session.
func Accept(w SetupResponseWriter, r *SetupRequest, mux *TrackMux) (*Session, error) {
	if rsp, ok := w.(*responseWriter); ok {
		return rsp.accept(mux)
	} else {
		return nil, fmt.Errorf("moq: invalid response writer type %T", w)
	}
}

File: moq-web/src/session_stream.ts
import { type CancelCauseFunc, type Context, withCancelCause } from "@okudai/golikejs/context";
import type { Stream } from "./internal/webtransport/mod.ts";
import { SessionUpdateMessage } from "./internal/message/mod.ts";
import type { SessionClientMessage, SessionServerMessage } from "./internal/message/mod.ts";
import { Cond, Mutex } from "@okudai/golikejs/sync";
import type { Version } from "./version.ts";
import { Extensions } from "./extensions.ts";

interface SessionStreamInit {
	context: Context;
	stream: Stream;
	client: SessionClientMessage;
	server: SessionServerMessage;
	detectFunc: () => Promise<number>;
}

export class SessionStream {
	#stream: Stream;
	readonly context: Context;
	#cancelFunc: CancelCauseFunc;
	#mu: Mutex = new Mutex();
	#cond: Cond = new Cond(this.#mu);
	#clientInfo: ClientInfo;
	#serverInfo: ServerInfo;

	#detectFunc: () => Promise<number>;
	#wg: Promise<void>[] = [];

	constructor(init: SessionStreamInit) {
		this.#stream = init.stream;
		this.#clientInfo = {
			versions: init.client.versions,
			extensions: new Extensions(init.client.extensions),
			bitrate: 0,
		};
		this.#serverInfo = {
			version: init.server.version,
			extensions: new Extensions(init.server.extensions),
			bitrate: 0,
		};
		[this.context, this.#cancelFunc] = withCancelCause(init.context);
		this.#detectFunc = init.detectFunc;

		// Cancel streams when context is cancelled
		this.context.done().then(() => {
			this.#stream.readable.cancel(0).catch(() => {});
			this.#stream.writable.cancel(0).catch(() => {});
		});

		// Start handling session updates
		this.#wg.push(
			this.#handleUpdates().catch((err) => {
				console.error(`moq: error in handleUpdates: ${err}`);
			})
		);

		// Start detecting bitrate updates
		// this.#backgroundTasks.push(
		// 	this.#detectUpdates().catch((err) => {
		// 		console.error(`moq: error in detectUpdates: ${err}`);
		// 	})
		// );
	}

	// async #detectUpdates(): Promise<void> {
	// 	while (!this.context.err()) {
	// 		const bitrate = await this.#detectFunc();
	// 		if (this.context.err()) {
	// 			break;
	// 		}
	// 		await this.#update(bitrate);
	// 	}
	// }

	async #handleUpdates(): Promise<void> {
		while (!this.context.err()) {
			const msg = new SessionUpdateMessage({});
			const err = await msg.decode(this.#stream.readable);
			if (err) {
				this.#cancelFunc(new Error(`moq: failed to decode session update message: ${err}`));
				break;
			}

			this.#serverInfo.bitrate = msg.bitrate;
			this.#cond.broadcast();
		}
	}

	// #update sends a session update message to the server.
	// It updates the client's bitrate and notifies the server of significant changes.
	// The bitrate should be originated from the WebTransport API.
	// TODO: get bitrate from WebTransport API and detect significant changes.
	async #update(bitrate: number): Promise<void> {
		const msg = new SessionUpdateMessage({ bitrate });
		const err = await msg.encode(this.#stream.writable);
		if (err) {
			this.#cancelFunc(new Error(`moq: failed to encode session update message: ${err}`));
			return;
		}

		this.#clientInfo.bitrate = msg.bitrate;

		return;
	}

	async updated(): Promise<void> {
		await this.#cond.wait();
	}

	get clientInfo(): ClientInfo {
		return this.#clientInfo;
	}

	get serverInfo(): ServerInfo {
		return this.#serverInfo;
	}

	async waitForBackgroundTasks(): Promise<void> {
		await Promise.allSettled(this.#wg);
	}
}

type ClientInfo = {
	versions: Set<Version>;
	extensions: Extensions;
	bitrate: number;
};

type ServerInfo = {
	version: Version;
	extensions: Extensions;
	bitrate: number;
};
